mimxrt: Upgrade the NXP library to version 2.10.

robert-hh opened the pull request:
The previous version of the library is v2.6. The actual version of the NXP library is v2.11. v2.10 is the first version that supports the i.MX RT 117x series. The different library versions are not fully compatible. I've chosen v2.10 because it is more close to the existing code base than v2.11. The switch from v2.10 to v2.11 requires a major rework of the clock configuration and can be made later.

The change to v2.10 allows for/requires some changes in the code:

- remove some part of pwm_backlog.*, which is provided by the lib now.
- change eth.c: The newer versions have additional parameters of
  the library versions.
- change sdcard.c: Use TransferBlocking instead of TransferNonblocking.
- add some support for the MIMXRT117x devices. More will come with the
  PR add the MIMXRT1176 board.
- Change the clock definiton for UART, I2C, Timer.
- Use blocking transfer only for SPI.
  It's faster and interferes less with other modules.
- Use the clock_config.c files of library v2.8.5
  The mimxrt port keeps the clock_config.c files from Version 2.8.5 and earlier.
  With clock_config.c from v2.10, the boards do not work. Refactoring of
  the clock set-up is on the to-do list.
- Enable expiry timers for UART, I2C and SPI, avoiding a stall in
  library code.

Reformat carried-over library code from NXP with uncrustify.

robert-hh committed:
mimxrt: Update nxp_driver to v2.10.

That allows for/requires some changes in the code:

- remove some part of pwm_backlog.*, which is provided by the lib now.
- change eth.c: The newer versions have additional parameters of
  the library versions.
- change sdcard.c: Use TransferBlocking instead of TransferNonblocking.
- add some support for the MIMXRT1176 device.
  The latter two changes were initially made for the mimxrt1176 port,
  but work with this library version as well.
- Set the clocks for UART, I2C, Timer.
- Integrate the I2S module and fix a rebase error.
- Use blocking transfer only for SPI.
  It's faster and interferes less with other modules.
- Use the clock_config.c files of library v2.8.5
  The mimxrt files keeps the clock_config.c files from Verson 2.8.5.
  With clock_config.c from v2.10, the boards do not work. Refactoring of
  the clock set-up is on the to-do list.
- Enable expiry timers for UART, I2C and SPI, avoiding a stall in
  library code.

robert-hh committed:
mimxrt: Format with uncrustify.

dpgeorge commented:
Thanks for the work, this looks like a good update.

Is it possible to stick with the hathach nxp_driver submodule?  I see there is a commit there with v2.10.

With the new clock_config.c files, are any of them duplicated, eg the TEENSY40 and TEENSY41?  If so it might be good to reuse the code.  Eg put the clock config files in `boards/clock_config_1062.c` and then reuse it in all 1062 boards.

robert-hh commented:
> Is it possible to stick with the hathach nxp_driver submodule? I see there is a commit there with v2.10.

I had asked hatach to upgrade his lib to include v2.10 and v2.11, which he kindly did. The port should point to these versions, at least that's the intention.

I can move the clock_config files to MCU specific copies in boards.

dpgeorge commented:
> The port should point to these versions, at least that's the intention.

Please try to change the submodule back to hathach, and point to the v2.10 commit in that repo.

robert-hh committed:
mimxrt: Make the clock_config.* files MCU-specific.

The clock_config.* files are moved from the board-specific directories
to the boards directory and given a MCU related name. The related
include statements in the source files had to be changed as well.

There is no functional change in this commit. All builds succeed.

robert-hh commented:
I changed the submodule URL and moved the clock_config.* files. All ports build fine, and I tested it with all boards that I have here.

dpgeorge commented on ports/mimxrt/Makefile:
```diff
@@ -66,6 +66,7 @@ CFLAGS += -Wno-error=unused-parameter
 
 INC += -I.
 INC += -Ihal
+INC += -Iboards
```
you won't need this line if you use `-DCLOCK_CONFIG_H='<boards/$(MCU_SERIES)_clock_config.h>'` below

robert-hh replied:
OK.

dpgeorge commented:
Please can you try to redo the commits so there is a separate commit (first) which just updates the NXP library version, then another commit which updates the mimxrt port to work with this new library.

robert-hh commented:
Isn't that already the case. There is:

f61d83bd1 mimxrt: Update nxp_driver to v2.10.
24e8c732f mimxrt: Format with uncrustify.
65344c9e5 mimxrt: Set the  nxp-driver submodule back to hatach's URL.
4b9be0370 mimxrt: Make the clock_config.* files MCU-specific.

Or do you suggest to split f61d83bd1 into several commits? I'm not sure if I can manage that.

robert-hh committed:
mimxrt: Implement a review comment and a uncrustify complain.

dpgeorge commented:
The "mimxrt: Update nxp_driver to v2.10." commit does most of the work, it updates the submodule and then changes the code.  I like to keep updates to submodules in distinct and separate commits, so they can be easily cherry-picked/reverted/rebased/etc.

I can do the adjustment to the commits during merge, if you like.

robert-hh commented:
> I can do the adjustment to the commits during merge, if you like.

I understand the intention, but I'm anxious to wreck the whole piece. So I'm happy if you could to that change.

dpgeorge commented:
> Looking into that commit, it is hard to split it. Just taking that single change out, that points to a different branch in the nxp-driver lib would result in an intermediate version that does not compile and does not run.

That's OK, that's how it's done with updates to `lib/stm32lib` (a follow up commit to make stm32 build against the new lib).

dpgeorge commented:
I split this into 2 commits and merged in dbe8b0c0ca20d6610c6d2b9b760485a55cacd72d and a0432ed9cb1a786db1e52ca1b3224d479c6f5f22

robert-hh commented:
Thank you very much.
